### PR TITLE
Add colours to pretty printing of values.

### PR DESCRIPTION
**Pull request by jameshaydon**

### CODE REVIEW

File: src/Radicle/Internal/Pretty.hs [hunk 1]
```diff
@@ -17,43 +20,49 @@ import qualified Radicle.Internal.Annotation as Ann
 import           Radicle.Internal.Core
 import           Radicle.Internal.Type
 
+class PrettyV a where
+  prettyV :: a -> Doc Type
```
Comment (jkarni) [inline src/Radicle/Internal/Pretty.hs]:
```suggestion
    prettyV :: a -> Doc Type
```

File: src/Radicle/Internal/Pretty.hs [hunk 2]
```diff
@@ -17,43 +20,49 @@ import qualified Radicle.Internal.Annotation as Ann
...
-        Atom i -> pretty i
-        Keyword i -> ":" <> pretty i
-        Ref i -> pretty i
-        String t -> "\"" <> pretty (escapeStr t) <> "\""
-        Number (a :% b) -> pretty a <> if b == 1 then "" else "/" <> pretty b
-        Boolean True -> "#t"
-        Boolean False -> "#f"
-        List vs -> case vs of
+instance PrettyV Type where
+  prettyV = prettyV . typeToValue
```
Comment (jkarni) [inline src/Radicle/Internal/Pretty.hs]:
```suggestion
    prettyV = prettyV . typeToValue
```

File: src/Radicle/Internal/Pretty.hs [hunk 3]
```diff
@@ -119,14 +128,26 @@ renderCompactPretty = renderStrict . layoutCompact . pretty
...
+toAnsi :: Type -> Term.AnsiStyle
+toAnsi = \case
+  TKeyword -> color Magenta
+  TString -> color Green
+  TNumber -> color Yellow
+  TBoolean -> color Cyan
+  TRef -> color Red
+  _ -> mempty
+
+renderAnsi :: PrettyV v => v -> Text
```
Comment (jkarni) [inline src/Radicle/Internal/Pretty.hs]: Could you put a doc/doctest as with `renderPrettyDef` above?

File: src/Radicle/Internal/Pretty.hs [hunk 4]
```diff
@@ -17,43 +20,49 @@ import qualified Radicle.Internal.Annotation as Ann
...
-        String t -> "\"" <> pretty (escapeStr t) <> "\""
-        Number (a :% b) -> pretty a <> if b == 1 then "" else "/" <> pretty b
-        Boolean True -> "#t"
-        Boolean False -> "#f"
-        List vs -> case vs of
+instance PrettyV Type where
+  prettyV = prettyV . typeToValue
+
+instance forall t. (Copointed t, Ann.Annotation t) => PrettyV (Ann.Annotated t ValueF) where
+    prettyV v = case v of
```
Comment (jkarni) [inline src/Radicle/Internal/Pretty.hs]: Can this be:
```suggestion
    prettyV v = annotate (typeOf v) $ case v of
```

And then no further (top-level) `annotates`?

File: src/Radicle/Internal/Pretty.hs [hunk 5]
```diff
@@ -17,43 +20,49 @@ import qualified Radicle.Internal.Annotation as Ann
...
+            sep [ prettyV k <+> prettyV val
                 | (k, val) <- Map.toList mp ]
-        Lambda ids vals _ -> parens $
-            "fn" <+> align (sep
-                        [ brackets . sep $ pretty <$> ids
-                        , sep $ pretty <$> toList vals
-                        ])
+        Lambda ids vals _ ->
+          let args :: Ann.Annotated t ValueF = Vec $ Seq.fromList (Atom <$> ids) in
+          annotate TFunction $ parens $
```
Comment (jkarni) [inline src/Radicle/Internal/Pretty.hs]: This section is a little confusingly formatted. Can the `in` be put on the next line, and arguments to applications indented?

File: test/spec/Radicle/Tests.hs [hunk 6]
```diff
@@ -809,7 +809,7 @@ test_macros =
                     , ":quit"
                     , "x"
                     ]
-            output = [ "0" ]
+            output = [ ansi "0" ]
```
Comment (jkarni) [inline test/spec/Radicle/Tests.hs]: Maybe just redefine `@==>`?

File: test/spec/Radicle/Tests.hs [hunk 7]
```diff
@@ -849,3 +849,8 @@ runInRepl inp = do
                                  , "repl.rad" `T.isSuffixOf` name
                                  ]
         pure $ runTestWithFiles testBindings inp (Map.fromList srcMap) (fromJust replSrc)
+
+ansi :: Text -> Text
```
Comment (jkarni) [inline test/spec/Radicle/Tests.hs]: Maybe a test for `renderAnsi`?

File: test/spec/Radicle/Tests.hs [hunk 8]
```diff
@@ -849,3 +849,8 @@ runInRepl inp = do
...
+ansi t = case parseTest t of
+  Left _ -> panic "Error adding colour codes to value string: original string does not parse."
```
Comment (jkarni) [inline test/spec/Radicle/Tests.hs]:
```suggestion
    Left _ -> panic "Error adding colour codes to value string: original string does not parse."
```

File: test/spec/Radicle/Tests.hs [hunk 9]
```diff
@@ -849,3 +849,8 @@ runInRepl inp = do
...
+  Right v -> renderAnsi v
```
Comment (jkarni) [inline test/spec/Radicle/Tests.hs]:
```suggestion
    Right v -> renderAnsi v
```

File: src/Radicle/Internal/Pretty.hs [hunk 10]
```diff
@@ -119,14 +128,26 @@ renderCompactPretty = renderStrict . layoutCompact . pretty
...
 -- >>> renderPrettyDef (asValue (List [String "hi", String "there"]))
 -- "(\"hi\" \"there\")"
-renderPrettyDef :: Pretty v => v -> Text
-renderPrettyDef = renderStrict . layoutSmart defaultLayoutOptions . pretty
+renderPrettyDef :: PrettyV v => v -> Text
+renderPrettyDef = renderStrict . layoutSmart defaultLayoutOptions . prettyV
+
+toAnsi :: Type -> Term.AnsiStyle
+toAnsi = \case
+  TKeyword -> color Magenta
```
Comment (jkarni) [inline src/Radicle/Internal/Pretty.hs]: Four spaces

File: src/Radicle/Internal/Effects.hs [hunk 11]
```diff
@@ -82,7 +82,7 @@ replPrimFns = fromList $ allDocs $
       , "Pretty-prints a value."
       , \case
         [x] -> do
-            putStrS (renderPrettyDef x)
+            putStrS (renderAnsi x)
```
Comment (jkarni) [inline src/Radicle/Internal/Effects.hs]: I think this should stay as `renderPrettyDef`, since otherwise we can't read it back!

Comment (jameshaydon) [reply src/Radicle/Internal/Effects.hs]: We might add another primop for this, or options to this one, later.

**Pull request merged**